File: Assignment-04/Question-04/mpi_root.py
#!/usr/bin/env python2
###############################################################################
#
# Assignment 4, Question 4 solution using MPI in Python.
#
# Copyright (C) 2014, Jonathan Gillett (100437638)
# All rights reserved.
#
###############################################################################
import numpy as np
from mpi4py import MPI

# Define process 0 as MASTER
MASTER = 0
# Set the subinterval limit
INTERVAL = 10E-9
# Set the width for search space
WIDTH = 1E6


def f(x):
    return -2.0 + np.sin(x) + np.sin(x**2) + np.sin(x**3) + \
        np.sin(x**999) + np.sin(x**1000)


def master(n_proc, comm):
    """The master process divides the interval [0, 1] into subintervals."""
    # Divide the interval [0,1] into several subintervals for each process
    intervals = np.linspace(0, 1, n_proc)
    ANS = np.zeros(2, dtype=np.float64)  # The result and r value from processes
    SOL = np.ones(2, dtype=np.float64)  # The final smallest solution and r value
    n_sent = 0  # Number of processes evaluating subintervals

    # Assign to each task an endpoint of one of the subintervals at which the
    # function f should be evaluated
    for k in range(1, n_proc):
        # Send the initial subintervals and width to each process
        comm.Send(intervals[k-1:k+1], k, tag=WIDTH)
        n_sent += 1

    while n_sent > 0:
        status = MPI.Status()
        # Receive the r value and result that closest approximates function f(x)
        comm.Recv(ANS, source=MPI.ANY_SOURCE, tag=MPI.ANY_TAG, status=status)
        sender = status.source
        width = status.tag

        # Record the results if solution is closer to zero
        print "ANS: %g, SOL: %g" % (ANS[0], SOL[0])
        if np.absolute(ANS[0]) <= np.absolute(SOL[0]):
            SOL = np.copy(ANS)

        # Terminate when the subinterval width is reduced below 10E-11
        if ((1.0 / n_proc) / width <= INTERVAL):
            comm.Send(intervals[sender-1:sender+1], sender, tag=0)
            n_sent -= 1
        else:
            # Divide subinterval into smaller pieces for execution again
            comm.Send(intervals[sender-1:sender+1], sender, tag=width*2.0)

    # Master process prints the approximate zero of the function f to standard output
    print "The approximate solution: %1.15g, for r = %1.15g" % (SOL[0], SOL[1])


def slave(proc_id, comm):
    """The slave process receives the intervals to calculate and the width."""
    bounds = np.zeros(2, dtype=np.float64)
    ans = np.zeros(2, dtype=np.float)
    status = MPI.Status()
    comm.Recv(bounds, source=MASTER, tag=MPI.ANY_TAG, status=status)
    width = status.tag

    while width > 0:
        # Analyse the function values determined by each process to find a new
        # subinterval [a,b]c[0,1] within which the solution r is guaranteed to lie.
        r = np.linspace(bounds[0], bounds[1], width)
        results = f(r)  # Calculate the result of the function for each value of r

        # Find the result that is closest to 0 and return the r value
        min_idx = np.argmin(np.absolute(results))
        ans[0], ans[1] = results[min_idx], r[min_idx]

        # Return the results to master process and wait for more work
        comm.Send(ans, MASTER, tag=width)

        # Clean up memory used
        del(r)
        del(results)

        status = MPI.Status()
        comm.Recv(bounds, source=MASTER, tag=MPI.ANY_TAG, status=status)
        width = status.tag


if __name__ == '__main__':
    comm = MPI.COMM_WORLD
    proc_id = comm.Get_rank()
    n_proc = comm.Get_size()

    if proc_id == MASTER:
        master(n_proc, comm)
    else:
        slave(proc_id, comm)

File: Assignment-04/Question-04/mpi_root.c
/*
 * Assignment 4, Question 4 solution using MPI in C.
 *
 * Copyright (C) 2015, Jonathan Gillett (100437638)
 * All rights reserved.
 */
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <mpi.h>

// Define process 0 as MASTER
#define MASTER 0
// Define the subinterval limit
#define INTERVAL 10E-11
// Set the width for the search space of each subinterval
#define WIDTH 1E7
// Message tags
#define REQUEST 1
#define REPLY   2


double f(double x) {
    return -2.0 + sin(x) + sin(pow(x, 2.0)) + sin(pow(x, 3.0)) + 
           sin(pow(x, 999.0)) + sin(pow(x, 1000.0)); 
}


// Sets the intervals given the bounds and number of subintervals
void set_intervals(double intervals[], double left, double right, int n) {
    double h = (right - left) / (double)(n-1);
    for (int i = 0; i < n; ++i) {
        intervals[i] = left + h*i;
    }
}


// The master process divides the interval [0, 1] into subintervals
void master(int n_proc) {
    double *intervals = (double *) malloc(n_proc*sizeof(double));
    // Divide the interval [0,1] into several subintervals for each process
    set_intervals(intervals, 0.0, 1.0, n_proc);
    double ANS[4] = { 0.0 };  // The result, r value, bounds from processes
    double SOL[4] = { 1.0 };  // The final smallest solution, r value, bounds
    int n_sent = 0;
    MPI_Status status;

    // Assign to each task an endpoint of one of the subintervals at which
    // the function f should be evaluated
    // Print intervals
    printf("INTERVALS: [%f, %f, %f, %f]\n", 
           intervals[0], intervals[1], intervals[2], intervals[3]);
    for (int i = 1; i < n_proc; ++i) {
        double bounds[2] = { intervals[i-1], intervals[i] };
        MPI_Send(bounds, 2, MPI_DOUBLE, i, REQUEST, MPI_COMM_WORLD);
        n_sent++;
    }

    while (n_sent > 0) {
        // Receive the r value and result that closest approximates function f(x)
        MPI_Recv(ANS, 4, MPI_DOUBLE, MPI_ANY_SOURCE, REPLY,
                 MPI_COMM_WORLD, &status);

        // Analyse the function values determined by each process to find a new
        // subinterval [a,b]c[0,1] within the solution r is guaranteed to lie
        printf("ANS: %1.15g, %f  SOL: %1.15g, %f\n", ANS[0], ANS[1], SOL[0], SOL[1]);
        if (fabs(ANS[0]) <= fabs(SOL[0])) {
            printf("NEW ANS: %1.15g, R: %1.15g  SOL: %1.15g, R: %1.15g\n",
                    ANS[0], ANS[1], SOL[0], SOL[1]);
            printf("NEW BOUNDS: [%1.15g, %1.15g]\n", ANS[2], ANS[3]);
            printf("OLD INTERVALS: [%f, %f, %f, %f]\n", 
                   intervals[0], intervals[1], intervals[2], intervals[3]);
            SOL[0] = ANS[0];
            SOL[1] = ANS[1];
            SOL[2] = ANS[2];
            SOL[3] = ANS[3];
            // Update the subintervals for the search space
            set_intervals(intervals, ANS[2], ANS[3], n_proc);
            printf("NEW INTERVALS: [%f, %f, %f, %f]\n", 
                   intervals[0], intervals[1], intervals[2], intervals[3]);
        }

        // Reduce num sent when below 10E-11
        if (((ANS[3] - ANS[2]) / WIDTH) <= INTERVAL) {
            n_sent--;
        }
        // Divide subinterval into smaller pieces for execution again
        double bounds[2] = { intervals[status.MPI_SOURCE-1],
                             intervals[status.MPI_SOURCE] };
        MPI_Send(bounds, 2, MPI_DOUBLE, status.MPI_SOURCE, REQUEST, MPI_COMM_WORLD);
    }
    // Terminate all processes
    for (int i = 1; i < n_proc; ++i) {
        double bounds[2] = { -1.0, -1.0 };
        MPI_Send(bounds, 2, MPI_DOUBLE, i, REQUEST, MPI_COMM_WORLD);
    }
    free(intervals);
    printf("Approx zero of function f(x): %1.15g\nr value: %1.15g\n", SOL[0], SOL[1]);
}


// The slave process receives the intervals to calculate and the width
void slave(int proc_id) {
    double bounds[2] = { 0.0 };
    double ans[4] = { 1.0 };
    double h = 0.0, r = 0.0, result = 0.0;
    MPI_Status status;
    MPI_Recv(bounds, 2, MPI_DOUBLE, MASTER, REQUEST,
             MPI_COMM_WORLD, &status);
    printf("BOUNDS: [%f, %f]\n", bounds[0], bounds[1]);
    while (bounds[0] >= 0) {
        h = (bounds[1] - bounds[0]) / (double)(WIDTH -1);
        // Given the width iterate through each value of r
        // to find the result that is closest to 0
        for (int i = 0; i < WIDTH; ++i) {
            r = bounds[0] + i*h;
            result = f(r);
            if (fabs(ans[0]) > fabs(result)) {
                ans[0] = result;
                ans[1] = r;
                ans[2] = bounds[0];
                ans[3] = bounds[1];
            }
        }
        // Return the results to master process and wait for more work
        MPI_Send(ans, 4, MPI_DOUBLE, MASTER, REPLY, 
                 MPI_COMM_WORLD);
        MPI_Recv(bounds, 2, MPI_DOUBLE, MASTER, REQUEST, 
                 MPI_COMM_WORLD, &status);
        printf("BOUNDS: [%f, %f]\n", bounds[0], bounds[1]);
    }
}


int main(int argc, char *argv[]) {
    int    proc_id;      // Process ID number
    int    n_proc;       // Number of processes

    // Initialize MPI
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &n_proc);
    MPI_Comm_rank(MPI_COMM_WORLD, &proc_id);

    if (proc_id == MASTER) {
        master(n_proc);
    }
    else {
        slave(proc_id);
    }

    MPI_Finalize();
    return 0;
}
